### PR TITLE
Create an arpack-ng-config.cmake that propagates arpack's dependencies

### PR DESCRIPTION
**Pull request by juanjosegarciaripoll**
## Pull request purpose

The CMake files that are installed by arpack are non-standard and do not enumerate the library dependencies. This is critical where arpack is not built as a shared library (e.g. Windows), but more generally it is not considered good practice. This way, users can just use
```
target_add_libraries(myproject PRIVATE arpack)
```
instead of having themselves to build something like
```
find_package(BLAS REQUIRED)
[... code for older versions of cmake's findBlas ...]
find_package(LAPACK REQUIRED)
[... code for older versions of cmake's findBlas ...]
target_add_libraries(myproject PRIVATE arpack BLAS::BLAS LAPACK::LAPACK)
```

## Detailed changes proposed in this pull request

- Create an arpack-ng-targets.cmake file that relies on CMake to create proper targets
- Declare BLAS::BLAS and LAPACK::LAPACK and MPI::Fortran as public dependencies of arpack and parpack,
- Rewrite arpack-ng-config.cmake to find its dependencies using standard tools (find_dependency)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -665,8 +665,9 @@ if(ICB)
     target_compile_definitions(pyarpack PRIVATE PY_MAJOR_VERSION="3")
     set(pyarpack_HDR ${PROJECT_SOURCE_DIR}/ICB ${PROJECT_SOURCE_DIR}/EXAMPLES/MATRIX_MARKET ${PROJECT_SOURCE_DIR}/EXAMPLES/PYARPACK)
     target_include_directories(pyarpack PUBLIC ${pyarpack_HDR} ${EIGEN3_INCLUDE_DIR} ${Boost_INCLUDE_DIRS} ${PYTHON_INCLUDE_DIRS})
-    target_link_libraries(pyarpack BLAS::BLAS LAPACK::LAPACK ${Boost_LIBRARIES} ${PYTHON_LIBRARIES})
+    target_link_libraries(pyPRIVATE PUBLIC arpack ${Boost_LIBRARIES} ${PYTHON_LIBRARIES})
```
Comment (mmuetzel) [inline CMakeLists.txt:668]: Is `pyPRIVATE` a typo?

Comment (juanjosegarciaripoll) [reply CMakeLists.txt:668]: Yes, sorry. I will fix it.

Comment (juanjosegarciaripoll) [reply CMakeLists.txt]: Done

**Comment by fghoussen** [comment]
@juanjosegarciaripoll : can you describe in detail what kind of problem you encounter on Window? Don't get it

**Comment by juanjosegarciaripoll** [comment]
> @juanjosegarciaripoll : can you describe in detail what kind of problem you encounter on Window? Don't get it

It's not exclusive of Windows but of any platform where the library is built statically and CMake is used to link to Arpack. In those situations the user must explicitly list not only Arpack but also all the blas, lapack and pthreads libraries used by Arpack, because the CMake exported files do not record this information (they only recorded the CMake equivalent of -larpack, which is not enough for statically linked libraries)

Moreover, the way that CMake files must be written is more or less how I have done it: the project must create a config file that reproduces the steps to find the libraries, not record the libraries as the pkgconfig files do. This is good practice and helps for the project to be easily relocatable.

So summing up, the CMake project needs this fix for the library to be used with CMake without any hacks. This fix is also required for the library to be easy and robustly integrated in other non-Unix, CMake-basef distribution such as vcpkg or Conan.

File: arpack-ng-config.cmake.in [hunk 2]
```diff
@@ -17,9 +17,43 @@ set(exec_prefix "@exec_prefix@")
...
+#set_target_properties(ARPACK::ARPACK PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${includedir}/arpack")
+#set_target_properties(ARPACK::ARPACK PROPERTIES INTERFACE_LINK_LIBRARIES      "arpack")
+#add_library(PARPACK::PARPACK INTERFACE IMPORTED)
+#set_target_properties(PARPACK::PARPACK PROPERTIES INTERFACE_LINK_LIBRARIES    "parpack")
+
+# Add the targets file
+include("${CMAKE_CURRENT_LIST_DIR}/arpack-ng-targets.cmake")
+add_library(ARPACK::ARPACK ALIAS arpack::arpack)
+if (@MPI@)
+    dd_library(PARPACK::PARPACK ALIAS arpack::parpack)
```
Comment (fghoussen) [inline arpack-ng-config.cmake.in:58]: Typo? add_library? Did you test that: does this works?

File: CMakeLists.txt [hunk 3]
```diff
@@ -637,20 +637,20 @@ if(ICB)
...
+  target_link_libraries(icb_arpack_cpp PRIVATE arpack ${EXTRA_LDFLAGS})
   add_test(icb_arpack_cpp_tst ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/icb_arpack_cpp)
 
   if (ICBEXMM)
     set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/EXAMPLES/MATRIX_MARKET)
 
     add_executable(arpackmm EXAMPLES/MATRIX_MARKET/arpackmm.cpp)
     target_include_directories(arpackmm PUBLIC ${PROJECT_SOURCE_DIR}/ICB ${EIGEN3_INCLUDE_DIR}) # Get arpack.h + eigen
-    target_link_libraries(arpackmm arpack BLAS::BLAS LAPACK::LAPACK ${EXTRA_LDFLAGS})
+    target_link_libraries(arpackmm PUBLIC arpack ${EXTRA_LDFLAGS})
```
Comment (fghoussen) [inline CMakeLists.txt:653]: Should this be PRIVATE?

File: CMakeLists.txt [hunk 4]
```diff
@@ -665,8 +665,9 @@ if(ICB)
     target_compile_definitions(pyarpack PRIVATE PY_MAJOR_VERSION="3")
     set(pyarpack_HDR ${PROJECT_SOURCE_DIR}/ICB ${PROJECT_SOURCE_DIR}/EXAMPLES/MATRIX_MARKET ${PROJECT_SOURCE_DIR}/EXAMPLES/PYARPACK)
     target_include_directories(pyarpack PUBLIC ${pyarpack_HDR} ${EIGEN3_INCLUDE_DIR} ${Boost_INCLUDE_DIRS} ${PYTHON_INCLUDE_DIRS})
-    target_link_libraries(pyarpack BLAS::BLAS LAPACK::LAPACK ${Boost_LIBRARIES} ${PYTHON_LIBRARIES})
+    target_link_libraries(pyarpack PUBLIC arpack ${Boost_LIBRARIES} ${PYTHON_LIBRARIES})
```
Comment (fghoussen) [inline CMakeLists.txt:668]: Should this be PRIVATE?

File: arpack-ng-config.cmake.in [hunk 5]
```diff
@@ -17,9 +17,43 @@ set(exec_prefix "@exec_prefix@")
...
+endif()
+if (NOT TARGET LAPACK::LAPACK)
+    find_dependency(LAPACK REQUIRED)
+    if (NOT TARGET LAPACK::LAPACK) # Create target "at hand" to ensure compatibility if cmake version < 3.18
+        add_library(LAPACK::LAPACK INTERFACE IMPORTED)
+        set_target_properties(LAPACK::LAPACK PROPERTIES INTERFACE_LINK_LIBRARIES "${LAPACK_LIBRARIES}")
+    endif()
+endif()
+if (@MPI@)
+    if (NOT TARGET MPI::MPI_Fortran) # Search only if not already found by upper CMakeLists.txt
```
Comment (fghoussen) [inline arpack-ng-config.cmake.in:36]: Missing MPI:MPI_C and MPI::MPI_CXX in case of ICB? https://github.com/opencollab/arpack-ng/blob/645db9e6472283408563e3e2dc99c5deb8272c5d/CMakeLists.txt#L220

File: arpack-ng-config.cmake.in [hunk 6]
```diff
@@ -17,9 +17,43 @@ set(exec_prefix "@exec_prefix@")
 set(libdir "@libdir@")
 set(includedir "@includedir@")
 
+include(CMakeFindDependencyMacro)
```
Comment (fghoussen) [inline arpack-ng-config.cmake.in:20]: If `find_dependency` are done, sounds like one must kill `find_package` in CMakeLists.txt? https://github.com/opencollab/arpack-ng/blob/645db9e6472283408563e3e2dc99c5deb8272c5d/CMakeLists.txt#L222

Comment (juanjosegarciaripoll) [reply arpack-ng-config.cmake.in:20]: No, find_package() is still required in CMakeLists.txt to determine how the library is built and to build the executables and tests themselves. find_dependency() is only used in the installed copy of arpack and only when the library is searched for by the user.

File: CMakeLists.txt (hunk 4 above)
Comment (juanjosegarciaripoll) [reply CMakeLists.txt:668]: This choice is consistent with the choice for target_include_directories(pyarpack...)

File: arpack-ng-config.cmake.in (hunk 5 above)
Comment (juanjosegarciaripoll) [reply arpack-ng-config.cmake.in]: This has been fixed in a new commit.

**Comment by fghoussen** [comment]
@juanjosegarciaripoll: can you rebase?

**Comment by fghoussen** [comment]
@juanjosegarciaripoll: can you report your changes into arpack-ng/cmake/arpac**kn**g-config.cmake.in and rebase on top of (new) master?

arpack-ng/cmake/arpac**k-n**g-config.cmake.in had to be moved into arpack-ng/cmake/arpac**kn**g-config.cmake.in in  3ba4dbfbe908340512a2d0d731598d843462e531: this was needed to be able to test arpack-ng install, your rebase missed that move.

>I am not a git expert

You are becoming one! :D

File: CMakeLists.txt [hunk 7]
```diff
@@ -650,8 +650,9 @@ if(ICB)
...
     install(TARGETS pyarpack
+            EXPORT arpack-ng-targets
```
Comment (fghoussen) [inline CMakeLists.txt:655]: @juanjosegarciaripoll: can you explain a bit what are the benefits of using targets and EXPORT?

**Comment by fghoussen** [comment]
@juanjosegarciaripoll: can you report your changes into arpack-ng/cmake/arpackng-config.cmake.in and rebase?

**Comment by juanjosegarciaripoll** [comment]
Apologies, I am away for holidays and conferences and have not been able to
address this. I need to find a weekend to do the rebating.

**Comment by fghoussen** [comment]
@juanjosegarciaripoll: that's OK! No hurry :)

**Comment by fghoussen** [comment]
Can you report your changes into arpack-ng/cmake/arpackng-config.cmake.in and rebase? (no merge). Strange message in PR at my side.

image

**Comment by juanjosegarciaripoll** [comment]
I haven't submitted a PR recently. I am working on my GitHub fork merging
upstream changes and trying to fix what is left and makes current CMake
build non-functional.

For instance, just discovered that there are multiple naming
inconsistencies between the directories used for installation, the name of
the CMake package and the name of the config files.

What is the desired name of the project? arpack, arpack-ng, arpackng...

**Comment by fghoussen** [comment]
> What is the desired name of the project? arpack, arpack-ng, arpackng...

The project is named arpack-ng (arpack is the original reference codebase - not maintained).

But sometimes one need to use arpackng (_without_ dash): cmake hints (`find_package` and `*.cmake` files mechanism rely on the default `XXX_DIR` env variable), but, AFAIR, bash doesn't allow variable names like `arpack-ng_DIR` (dash _not_ allowed in bash variable names) so I had to rename the cmake file into arpackng-config.cmake.in (_without_ the dash) to get a way to run the test `tstCMakeInstall` OK: https://github.com/opencollab/arpack-ng/blob/8ecf1be71347f6b6eff5fc4f8c48e509cf3f4add/cmake/tstCMakeInstall.sh.in#L31

Please, do the minimal impact, I don't want to break everything: do as less as possible! :)

**Comment by juanjosegarciaripoll** [comment]
I have made various tests with Debian bullseye and CMake 3.18. The problem is that the current CMakeLists.txt installs the configuration files under the `arpack-ng` directory. In that situation `cmake` does not seem able to find the package. However, in the current rebased branch I changed CMakeLists.txt to install files under `lib/cmake/arpackng` and now CMake finds the configuration file. I am using the tests in this repository to verify the CMake configuration is properly installed and usable for C sources https://github.com/juanjosegarciaripoll/arpack-ng-cmake-test

**Comment by fghoussen** [comment]
@juanjosegarciaripoll: thanks! I had no time to have a close look (maybe this week-end), but, 2 on-the-fly remarks:

- there should be only one `.cmake` file, but, seems there are 2: `arpack-ng-config.cmake.in` (`arpack-ng` _with_ dash) and `./cmake/arpackng-config.cmake.in` (`arpackng` _without_ dash). My understanding is that the one that should stay is `./cmake/arpackng-config.cmake.in` (for reasons explained above) and that modifications done in `arpack-ng-config.cmake.in`  should have been done in  `./cmake/arpackng-config.cmake.in`
- autotools should do  the same as cmake: can you change the installation directory (libraries and include) from `arpack-ng` to `arpackng` for both autools (Makefile.am, ...) and cmake? (to get consistent install whatever the build system used)

**Comment by juanjosegarciaripoll** [comment]
You were right. I have also noticed that the current configuration did not propagate dependencies when the libraries were built in statically linked mode. I am working on further changes, as well as some cleanup of the code that builds the tests and examples.

**Comment by fghoussen** [comment]
This patch is becoming huge which is no good: mixing-up things is the best way to reg! :(
I thought at first it was only about using `find_dependency` in `./cmake/arpackng-config.cmake.in` (arpackng _without_ dash): can you just stick to and push that please? Others things could come after: one minimal step at a time.

Note: seems your patch creates `./arpack-ng-config.cmake.in` (arpack-ng _with_ dash) which should not be used (or I missed something in the process - this PR has a blurry history!...).

**Comment by juanjosegarciaripoll** [comment]
I am afraid this cannot be a small patch. There are many errors in the
actual CMake file because of the lack of order in the construction:

- find_dependency is not used, but also...
- targets do not set the include paths properly so it does not suffice to
include it
- the configuration file did not use CMake targets export file, which is
required to carry on dependencies on BLAS, LAPACK, MPI
- GNUInstallPaths was included late, leading to variables being wrongly
defined
- tests and examples are defined before all targets they depend on are
completely constructed
- the library only worked with shared library linking, not statically
linked projects
- configure_file was used where a file copy was required, placing various
tests at risk
- tests use hardcoded locations for the executables which CMake does not
follow on some platforms
- at various places the script redefined the global variable that defines
the location of binary files, an effect that propagates in a risky way
throughout the script; the new version avoids this
- moreover, those hardcoded locations do not work in Windows, where the DLL
is not available in the executable path; for that reason none of the
examples work in mingw64 and only some of the tests worked (none of the
ICBEXMM) as confirmed by your GitHub workflow which does not include those
options
- the arpack .GitHub/workflow tests for CMake are themselves wrong because
they supply the build script with the location of the CMake file which does
not test the discovery of those files
...

The history may seem complicated but the fact is that the CMake file is not
functioning and requires a lot of fixes, many of which are coupled among
themselves. It has various intermediate fixes because I do not have access
to all test platforms (OSX) and these reveal further inconsistencies in the
original code.

Also, taking the tests and examples outside of the main source is good
practice and helped me debugging the actual main targets. Your autoconf
version also does this. This also avoids having adhoc rules that are broken
by having common standard function for all tests, and solves the problem of
execution order I mentioned before.

However, my intention is to clean up a single patch with a cleaner
description after rebasing my branch to a single commit.

Regarding your last remark the patch is not creating an arpack-ng file but
arpackng prefixed ones. I have tested it on Linux and Windows. Currently
the only platform that does not run all tests is OSX because I forgot a
$<TARGET_FILE> macro.

**Comment by juanjosegarciaripoll** [comment]
I am sorry if the patch is large, but I think once cleaned up it is not difficult to read, and solves important issues.

For instance, the Windows build now runs with all tests and examples, even in Mingw64. The CMake files no longer uses non-standard practices, such as setting the global directory for executable files.

If the history of this pull request is too messy, I apologize for it. If required, I can take this patch into a fresh new branch.

**Comment by juanjosegarciaripoll** [comment]
It seems that there is something wrong I am doing with the OS X platform. I have been able to refactor the dependencies tracking to a smaller patch which I am moving to a different pull request https://github.com/opencollab/arpack-ng/pull/377

**Comment by fghoussen** [comment]
> The history may seem complicated but the fact is that the CMake file is not functioning and requires a lot of fixes

I agree. Would be good if you could propose one (minimal) PR one-step-at-a-time for all bugs you listed above (if you have time for it?). We would be glad to merge them :) `arpack-ng` is made of volunteers (no pay...), mostly unix-based and shared lib I guess, trying to use cmake in the best possible way: not perfect but we are opened to improvements, so do not hesitate to PR fixes (one step-at-a-time is easier for maintainers to follow - I only stick around one or twice max a week... For free...)

> taking the tests and examples outside of the main source is good practice and helped me debugging the actual main targets. Your autoconf version also does this.

I'am OK with that. @sylvestre: are you? (I would say we have one only CMakeLists.txt for historical reasons).
Yes, autotools does this too. Doing the same with CMake would imply lots of files (CMakeLists.txt) in each subfolder.

> solves important issues.

I agree. Do not hesitate to PR next after #377 is merged

> those hardcoded locations do not work in Windows ... Windows build now runs with all tests and examples, even in Mingw64

Would be great if we could have all running OK on Windows too! :) AFAIR, we only have partial build and/or tests in  the CI.


> only platform that does not run all tests is OSX... something wrong I am doing with the OS X platform

A step-by-step PR approach will certainly help to understand what/why/when things break on OSX

**Comment by fghoussen** [comment]
> I am afraid this cannot be a small patch. There are many errors in the
actual CMake file because of the lack of order in the construction:

- find_dependency is not used, but also...
- targets do not set the include paths properly so it does not suffice to
include it
- the configuration file did not use CMake targets export file, which is
required to carry on dependencies on BLAS, LAPACK, MPI
- GNUInstallPaths was included late, leading to variables being wrongly
defined
- tests and examples are defined before all targets they depend on are
completely constructed
- the library only worked with shared library linking, not statically
linked projects
- configure_file was used where a file copy was required, placing various
tests at risk
- tests use hardcoded locations for the executables which CMake does not
follow on some platforms
- at various places the script redefined the global variable that defines
the location of binary files, an effect that propagates in a risky way
throughout the script; the new version avoids this
- moreover, those hardcoded locations do not work in Windows, where the DLL
is not available in the executable path; for that reason none of the
examples work in mingw64 and only some of the tests worked (none of the
ICBEXMM) as confirmed by your GitHub workflow which does not include those
options
- the arpack .GitHub/workflow tests for CMake are themselves wrong because
they supply the build script with the location of the CMake file which does
not test the discovery of those files
...


@juanjosegarciaripoll: do not hesitate  to PR other fix. Would be good to have full tests running on W10 in CI
